language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -228,43 +231,54 @@ typeOf (TyInst x t ty) = do
             k' <- kindOf ty
             typeCheckStep
             if k == k'
-                then pure (tyReduce (tySubstitute (extractUnique n) (void ty) ty''))
+                then tyReduce =<< tySubstitute (extractUnique n) (void ty) ty''
                 else throwError (KindMismatch x (void ty) k k')
         _ -> throwError (TypeMismatch x (void t) (TyForall () dummyTyName dummyKind dummyType) (void ty'))
 typeOf (Unwrap x t) = do
     ty <- typeOf t
     case ty of
         TyFix _ n ty' -> do
-            let subst = tySubstitute (extractUnique n) ty ty'
-            pure (tyReduce subst)
+            subst <- tySubstitute (extractUnique n) ty ty'
+            typeCheckStep
+            tyReduce subst
         _             -> throwError (TypeMismatch x (void t) (TyFix () dummyTyName dummyType) (void ty))
 typeOf (Wrap x n ty t) = do
     ty' <- typeOf t
-    let fixed = tySubstitute (extractUnique n) (TyFix () (void n) (void ty)) (void ty)
+    fixed <- tySubstitute (extractUnique n) (TyFix () (void n) (void ty)) (void ty)
     typeCheckStep
-    if tyReduce fixed == ty'
+    red <- tyReduce fixed
+    if red == ty'
         then pure (TyFix () (void n) (void ty))
         else throwError (TypeMismatch x (void t) fixed (void ty'))
 
 extractUnique :: TyNameWithKind a -> Unique
 extractUnique = nameUnique . unTyName . unTyNameWithKind
 
+fixUniversals :: MonadQuote m => Type TyNameWithKind a -> m (Type TyNameWithKind a)
```

Comment:
What does this do? Why do we need it? Can we not do this in a more principled way?

Reply:
Yes, probably. I wanted something Chad could use in the meantime, mainly.

Reply:
As for how it works: currently, when substituting in a value (which Chad noticed with `unwrap`), we substitute naivlely. If, however, we substitute a term containing universals, this can mean that the global uniqueness of names is not preserved, because we didn't generate a new bound variable.

Reply:
Please put that in the code.

On the code itself:
- What about type lambdas and fixpoints? Those both bind variables and so will suffer from the same problem when substituting.
- Perhaps it would be clearer to have a function called something like `freshenBoundVariables` that replaces all bound variables with fresh names? That's the general case, right?

Reply:
Also, I think we could probably construct some reduced test cases that illustrate precisely the issue, rather than being as large as Chad's example.